### PR TITLE
Sha ni

### PR DESCRIPTION
**Pull request by roconnor-blockstream**
Relax timing limit somewhat, and check for sha_ni CPU support.

Timing in testing require sha_ni CPU support.

### CODE REVIEW

File: C/cpu.c [hunk 1]
```diff
@@ -0,0 +1,191 @@
+#include "simplicity/cpu.h"
+
+#if defined(__SHA__) && defined(__SSE4_1__) && (defined(__x86_64__) || defined(__amd64__) || defined(__i386__))
+#define SHA_NI
+#endif
+
+#ifdef SHA_NI
+
+#include <cpuid.h>
```
Comment (uncomputable) [inline C/cpu.c:9]: 3563f344d22b59ce27af5c07ff7998259d742699: If any header is undefined on the target architecture then there will be an error, right? It seems that x86_64, amd64 and i386 all support cpuid, etc.

File: C/regression4.h [hunk 2]
```diff
@@ -0,0 +1,19 @@
+#ifndef SIMPLICITY_REGRESSION4_H
+#define SIMPLICITY_REGRESSION4_H
+
+#include <stddef.h>
+#include <stdint.h>
+#include "bounded.h"
+
+/* A length-prefixed encoding of the following Simplicity program:
+ *     uWitness OneV : f 15 ++ [uComp (3*2^16) 1]
```
Comment (uncomputable) [inline C/regression4.h:9]: This looks like the program in the OP of https://github.com/BlockstreamResearch/simplicity/issues/209, but with an additional `++ [uComp (3*2^16) 1]`?

**Review by uncomputable — APPROVED** [review]
ACK dceeaae127d9ab461ee470febeaae765c3cb0c9a Looks good and works on my end. I'm happy that these 9000 lines of changes stem from a single byte blob test vector. I was scared initially.

File: C/cpu.c [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+#endif
+
+/* WARNING: This function isn't threadsafe and must not be called concurrently with any function in this library, including itself.
+ * It is encouranged that you call this function once in the main function of your program
+ *
+ * This function inspects the CPU's capabilities and enables the use of some optimized instructions, if they are known and available.
+ *
+ * If the CPU supports Intel sha_ni operations, the result value has the SIMPLICITY_HAS_SHA_NI_FLAG bit set.
+ */
+uint32_t simplicity_cpu_optimize_not_thread_safe(void) {
```
Comment (uncomputable) [inline C/cpu.c:182]: AFAICS, SHA NI is not enabled by default if it is available. One has to call the `simplicity_cpu_optimize_not_thread_safe` function to explicitly opt into the optimization.

File: C/cpu.c (hunk 1 above)
Comment (roconnor-blockstream) [reply C/cpu.c:9]: `cpuid.h` is provided by gcc and clang.  It may just fail to exist on other compilers.

**Comment by apoelstra** [comment]
utACK dceeaae127d9ab461ee470febeaae765c3cb0c9a. Will hold off on running my test rig while you are playing with stuff.

**Comment by roconnor-blockstream** [comment]
So I f'd up the source branch of this PR.  I've made a new PR at #241.

I guess new comments should go there.

File: C/regression4.h (hunk 2 above)
Comment (roconnor-blockstream) [reply C/regression4.h:9]: the last uComp composes the main program with `uWitness`.

Here the `:` is list cons in Haskell, which always confuses me when I read this comment to myself.

**Pull request closed without merging**